packages/core/src/api/formatConversions/formatConversions.ts

```diff
@@ -1,133 +1,140 @@
+// // that outputs a data-language attribute instead of a CSS class (e.g.: language-typescript)
+// function code(state: any, node: any) {
+//   const value = node.value ? node.value + "\n" : "";
+//   /** @type {Properties} */
+//   const properties: any = {};
+//
+//   if (node.lang) {
+//     // changed line
+//     properties["data-language"] = node.lang;
+//   }
+//
+//   // Create `<code>`.
+//   /** @type {Element} */
+//   let result: any = {
+//     type: "element",
+//     tagName: "code",
+//     properties,
+//     children: [{ type: "text", value }],
+//   };
+//
+//   if (node.meta) {
+//     result.data = { meta: node.meta };
+//   }
+//
+//   state.patch(node, result);
+//   result = state.applyData(node, result);
+//
+//   // Create `<pre>`.
+//   result = {
+//     type: "element",
+//     tagName: "pre",
+//     properties: {},
+//     children: [result],
+//   };
+//   state.patch(node, result);
+//   return result;
+// }
+//
+// export async function markdownToBlocks<BSchema extends BlockSchema>(
+//   markdown: string,
+//   blockSchema: BSchema,
+//   schema: Schema
+// ): Promise<Block<BSchema>[]> {
+//   const htmlString = await unified()
+//     .use(remarkParse)
+//     .use(remarkGfm)
+//     .use(remarkRehype, {
+//       handlers: {
+//         ...(defaultHandlers as any),
+//         code,
+//       },
+//     })
+//     .use(rehypeStringify)
+//     .process(markdown);
+//
+//   return HTMLToBlocks(htmlString.value as string, blockSchema, schema);
+// }
+
+export async function markdown(cleanHTMLString: string) {
   const markdownString = await unified()
     .use(rehypeParse, { fragment: true })
     .use(removeUnderlines)
     .use(rehypeRemark)
     .use(remarkGfm)
     .use(remarkStringify)
-    .process(await blocksToHTML(blocks, schema));
+    .process(cleanHTMLString);
 
   return markdownString.value as string;
 }
-
-// modefied version of https://github.com/syntax-tree/mdast-util-to-hast/blob/main/lib/handlers/code.js
-// that outputs a data-language attribute instead of a CSS class (e.g.: language-typescript)
-function code(state: any, node: any) {
-  const value = node.value ? node.value + "\n" : "";
-  /** @type {Properties} */
-  const properties: any = {};
-
-  if (node.lang) {
-    // changed line
-    properties["data-language"] = node.lang;
-  }
-
-  // Create `<code>`.
-  /** @type {Element} */
-  let result: any = {
-    type: "element",
-    tagName: "code",
-    properties,
-    children: [{ type: "text", value }],
-  };
-
-  if (node.meta) {
-    result.data = { meta: node.meta };
-  }
-
-  state.patch(node, result);
-  result = state.applyData(node, result);
-
-  // Create `<pre>`.
-  result = {
-    type: "element",
-    tagName: "pre",
-    properties: {},
-    children: [result],
-  };
-  state.patch(node, result);
-  return result;
-}
-
-export async function markdownToBlocks<BSchema extends BlockSchema>(
```

Comment:
I understand it's temporarily commented out (similar to some stuff above), but we need to make a rollout plan:
- do an intermediate release with the new code for use-case 1+2, and the old code for markdown / html / etc
- don't do an intermediate release, release only until we have all use-cases solved. maybe still make a separate branch where we can squash use-case 1+2 into

Reply:
I think we should just roll everything out at once tbh. For one, we need to do S3 (paste from external) to finish fixing copy/paste. Once we do that I think it's largely a matter of extracting existing code into functions for the remaining scenarios, which shouldn't take long.